### ISSUE TITLE
[SP2019]Get Vision Working (libkoki)

### ISSUE THREAD

**Issue by jcwilliamson** [sourcebots/tasks#9]

**Comment by PeterJCLaw** [comment]
@jcwilliamson are there any more details on this?
Specifically:
- does this need to be `libkoki`? (if so, why?)
- what software platform is this for?
  - Python 2 or 3?
  - Raspbian or something else?
  - (Is it for Smallpeice?)
  - note for clarity: I'm exclusively asking about things _below_ the vision processing layer here, I don't think it matters what API/etc. sits on top of it
- what hardware does this related to?
   I'm assuming it would be for Raspberry Pis since that's the "brain" hardware that SourceBots has.
   SourceBots has some cameras (TeckNet ones I think), is the plan to use those? (if so it's worth noting that they've never been used with `libkoki` before)

**Comment by trickeydan** [comment]
>     * does this need to be `libkoki`? (if so, why?)

There are a large number of people that don't trust the stability of `sb-vision` and `libkoki` has been proven to work many time over. 

>     * what software platform is this for?

>       * Python 2 or 3?

3. I've determined that it shouldn't be too hard to port `pykoki`, as it is just a `ctypes` API. No need to upgrade OpenCV.

>       * Raspbian or something else?

All Linuxes with V4L2

>       * (Is it for Smallpeice?)

Yes. It would be good if we could actually ship vision this year, as imo it is one of the most impressive things from a competitor perspective.

>       * note for clarity: I'm exclusively asking about things _below_ the vision processing layer here, I don't think it matters what API/etc. sits on top of it
> 
>     * what hardware does this related to?
>       I'm assuming it would be for Raspberry Pis since that's the "brain" hardware that SourceBots has.

Correct. I'd like to have a working version of `libkoki` on a RasPi for SR purposes if we change brain board to  RasPi.

>       SourceBots has some cameras (TeckNet ones I think), is the plan to use those? (if so it's worth noting that they've never been used with `libkoki` before)

SourceBots has a variety of cameras, and we need to purchase more this year anyway. There were multiple issues with the TeckNet ones (sorry for being vague, I don't remember the specifics. cc @RealOrangeOne), so it's possible that we could switch to a different camera. Notably, HR Robocon has had success with Pi + Pi Camera v2 + `libkoki`.

Pi Camera specs for reference:
> The width of the sensor is about 3.66mm and focal length 3mm.
> 8 megapixel native resolution sensor-capable of 3280 x 2464 pixel static images
> Fixed focus lens
Robocon Vision Code: https://github.com/systemetric/robot/blob/master/robot/vision.py

**Comment by PeterJCLaw** [comment]
> > * does this need to be `libkoki`? (if so, why?)
> 
> There are a large number of people that don't trust the stability of `sb-vision` and `libkoki` has been proven to work many time over.

There are very few people who have actually looked at in depth the internals of either, let alone both. If there are technical reasons for choosing one over the other, please let's discuss them. We shouldn't be basing this on FUD.

Note that `libkoki` has a very narrow track record - a couple of camera and hardware combinations. April Tags, which `sb-vision` is based on, has a _much_ wider track record.

> > * what software platform is this for?
> >   * Python 2 or 3?

> 3. I've determined that it shouldn't be too hard to port `pykoki`, as it is just a `ctypes` API. No need to upgrade OpenCV.

I actually ended up looking at this yesterday (for other reasons), it turns out that `pykoki` depends on `cv` and `cv2` Pytjom imports at the moment. While I agree we don't want to update OpenCV if we can avoid it, I don't know if there exists a Python 3 + OpenCV 2 package.

Note also that currently `pykoki` depends on `v4l2`, a package which doesn't support Python 3 and appears unmaintained. This is fixable, but worth being aware of.

> > * (Is it for Smallpeice?)
> 
> Yes. It would be good if we could actually ship vision this year, as imo it is one of the most impressive things from a competitor perspective.

Bear in mind that our purpose at Smallpeice (and SR for that matter) is to provide an environment where the entrants can _learn_. It's not about impressing them. Not having vision previously has actually resulted in a wider variety of sensors being used, which implies a greater degree of learning happened.

As far as I recall, we _did_ ship vision last year (2018).

**Comment by trickeydan** [comment]
> Note that `libkoki` has a very narrow track record - a couple of camera and hardware combinations. April Tags, which `sb-vision` is based on, has a _much_ wider track record.

Fair point. However, `sb-vision` has been deployed once to my knowledge and had major thresholding issues.

> I actually ended up looking at this yesterday (for other reasons), it turns out that `pykoki` depends on `cv` and `cv2` Pytjom imports at the moment. While I agree we don't want to update OpenCV if we can avoid it, I don't know if there exists a Python 3 + OpenCV 2 package.

`cv` and `cv2` are only required if you want to pass a `IPLImage` directly into the library instead of using the `libkoki` drivers.

> Note also that currently `pykoki` depends on `v4l2`, a package which doesn't support Python 3 and appears unmaintained. This is fixable, but worth being aware of.

`v4l2` is only used for `v4l2.v4l2_format`, as `libkoki` has it's own `v4l2` driver. 

`v4l2_format` is a struct defined [here](https://github.com/AlexJinlei/python-v4l2/blob/8ac50e1220c310988cbb901e0f0db3fd8ec884e2/v4l2.py#L1810). We can easily move that across, the licences are compatible between the projects (`pykoki` isn't licensed yet, but will likely have to be GPL3, as `libkoki` is GPL3).

> Bear in mind that our purpose at Smallpeice (and SR for that matter) is to provide an environment where the entrants can _learn_. It's not about impressing them. Not having vision previously has actually resulted in a wider variety of sensors being used, which implies a greater degree of learning happened.

Given that the course scores fell dramatically last year, I think we need to try quite hard to improve the summer school this year, and _actually_ shipping vision is a good way to go about that in my opinion. Game Design should ensure that competitors use a variety of sensors, vision should be there to augment the game for those teams that have sufficient skills to use it.

> As far as I recall, we _did_ ship vision last year (2018).

Vision was definitely not shipped at Smallpeice last year, see photos of [arena](https://lockwood.me/blog/smallpeice-2018/).

**Comment by RealOrangeOne** [comment]
The technet cameras actually worked very well for what we needed them to. The only issue was that the mounting wasnt ideal (although it did have a screw mount which was a plus), and that it had manual focus, which required teams to calibrate and then fix the focus themselves. If there's budget, different cameras should be purchased. Alternatively, we can almost certainly use SRs pool of cameras, which are known to work well with `libkoki`.

I agree simply porting `libkoki` to python 3 (and changing *nothing* else) is the most efficient path. As an aside, I've been working on a vision system to replace `libkoki` and `sb-vision`, codename `yuri`, but I don't recommend shipping that at smallpeice, i'd like to use that as a testing ground for it before shipping.

RE whether we should ship vision, I personally feel vision definitely adds that _wow factor_ which courses like this benefit from greatly. I do however believe having something too flashy leads to a reliance on it. I suspect shipping vision in its usual form, but ensuring we have enough sensors which would work better (ultrasound is a good example) will not only give teams the option, but also make them think about (_learning_) the right tool for the job. Not shipping vision can also lead teams towards dead reckoning, something we want to avoid!

> > As far as I recall, we did ship vision last year (2018).

> Vision was definitely not shipped at Smallpeice last year, see photos of arena.

You're both correct, technically. We shipped vision last year at 'SourceBots 2018', but not at 'Smallpeice 2018'. Although that really doesn't matter.

**Comment by PeterJCLaw** [comment]
> > > As far as I recall, we did ship vision last year (2018).
> 
> > Vision was definitely not shipped at Smallpeice last year, see photos of arena.
> 
> You're both correct, technically. We shipped vision last year at 'SourceBots 2018', but not at 'Smallpeice 2018'. Although that really doesn't matter.

Nah, I'm probably wrong. For some reason I'd thought that we'd shipped the same vision from SB2018 at last year's Smallpeice. In retrospect I do recall us actually making the decision to remove it and hiding it from the docs.

**Comment by PeterJCLaw** [comment]
> > Bear in mind that our purpose at Smallpeice (and SR for that matter) is to provide an environment where the entrants can _learn_. It's not about impressing them. Not having vision previously has actually resulted in a wider variety of sensors being used, which implies a greater degree of learning happened.
> 
> Given that the course scores fell dramatically last year, I think we need to try quite hard to improve the summer school this year, and _actually_ shipping vision is a good way to go about that in my opinion. Game Design should ensure that competitors use a variety of sensors, vision should be there to augment the game for those teams that have sufficient skills to use it.

The drop in scores is interesting and somewhat unfortunate to hear (and news to me). I'll reiterate my point though -- the summer school is about learning, rather than being impressed with what someone else has done. Yes, it's _cool_ to have vision, however I think it's quite a stretch to say that that should be a primary focus based on negative feedback (especially as the 2017 summer school also didn't have vision, and did so in a rather spectacular manner).

I have thoughts on how we could address such feedback, though this probably isn't the right venue for them (though the core thesis would be to focus on making the actual learning experience good, rather than worrying about what specific technologies are used). Given our limited resources, if improving in light of that feedback is a focus, then adding a new piece of tech should definitely be out of scope.

**Comment by trickeydan** [comment]
> The drop in scores is interesting and somewhat unfortunate to hear (and news to me). 

You may be interested in viewing it. You should have access to the file in the SourceBots Google Drive.

It was heavily discussed last september after we received it.

**Comment by PeterJCLaw** [comment]
Aha, thanks. Is there a record of that discussion somewhere? I don't recall seeing anything on the mailing list. (Slack obviously doesn't count as a record, as much as it would want you to believe it is one it's nigh impossible to read old things in it). Either way I'm going to move this to https://github.com/sourcebots/tasks/issues/59.

**Comment by trickeydan** [comment]
> Aha, thanks. Is there a record of that discussion somewhere? I don't recall seeing anything on the mailing list. (Slack obviously doesn't count as a record, as much as it would want you to believe it is one it's nigh impossible to read old things in it). Either way I'm going to move this to #59.

It will be in the `minutes` archive I suspect.

**Comment by PeterJCLaw** [comment]
> It will be in the `minutes` archive I suspect.

Do you have a link for that? I've tried [searching GitHub](https://github.com/sourcebots?q=minutes) without luck.

**Comment by trickeydan** [comment]
> > It will be in the `minutes` archive I suspect.
> 
> Do you have a link for that? I've tried [searching GitHub](https://github.com/sourcebots?q=minutes) without luck.

https://github.com/s-r-o/minutes/blob/bfc77dc7ae8612dece848e118241c034683b08f0/2018/2018-11-26.md

**Comment by PeterJCLaw** [comment]
Some thoughts on the technical side of this:

A first priority needs to be to work out what hardware we're going to use. We've had considerable challenges getting cameras which work ok on Raspberry Pis. The TeckNet ones worked ok, but are just a bit crappy as cameras (not great resolution, not especially fast response times, not great optics, fixed focus). It's worth noting that we used them because the cameras which had been being used with the v3 kits caused USB resets when they were used (and that was reproducible using standalone command-line tooling). It's plausible that the reset issue has been fixed with more recent Raspbians, though I somewhat doubt it. Research here is definitely needed.

Obviously without known-working hardware any software efforts are redundant.

For libkoki, as @trickeydan suggests, I suspect that either munging `pykoki` so that it drops the direct dependencies on v4l2 and OpenCV are probably reasonable. Frankly `pykoki` is also _tiny_, so replacing it with something based on `cffi` rather than `ctypes` is viable (likely a day's effort for someone who's familiar with both `cffi` and `libkoki`, probably nearer two or more days for someone unfamiliar with them).

That would also provide a route to introduce a new API without worrying about breaking existing users which in turn would allow us to both make the dependencies on v4l2 and OpenCV more clearly separate and to add support for them _later_ (assuming we don't need them now). As its API surface is also tiny, I'm not worried about updating the current users (of which I'm aware of about three) to use the new library.
If we really wanted to, we could then even bind directly to v4l2 and opencv using our own `cffi`, to avoid needing to worry about their python layers. This feels maybe reasonable as we're using each only for a couple of data types, rather than anything more. I'm not sure if this is actually a good idea though.
